### PR TITLE
Integrate PollingController mixin with GasFeeController

### PR DESCRIPTION
**Pull request by adonesky1**
Integrates recently introduced  [`PollingController` mixin](https://github.com/MetaMask/core/pull/1736)  with `GasFeeController`. Leaves old polling pattern in pace for now so as to not force a ton of breaking changes for mobile, but with intention to activate new pattern in both clients ASAP.

Addresses: https://github.com/MetaMask/MetaMask-planning/issues/1314

### CODE REVIEW

File: packages/gas-fee-controller/src/GasFeeControllerPolling.ts
Comment (jiexi) [inline packages/gas-fee-controller/src/GasFeeControllerPolling.ts:47]: can we prepend the networkClientId to the polling token returned by `start()` or something to make it so that only pollingToken is required when calling this method?

File: packages/gas-fee-controller/src/GasFeeControllerPolling.ts
Comment (jiexi) [inline packages/gas-fee-controller/src/GasFeeControllerPolling.ts:69]: Some pollers may expose values that they update on each polling interval, but that they want to clear after some timeout from when they are stopped. See https://github.com/MetaMask/eth-block-tracker/blob/main/src/PollingBlockTracker.ts#L173

Maybe a `abstract pollStopped()`?

File: packages/gas-fee-controller/src/GasFeeControllerPolling.ts
Comment (BelfordZ) [inline packages/gas-fee-controller/src/GasFeeControllerPolling.ts:49]: Should we split this out into a stop & a seperate `stopAll` method, and tighten up this interface?

Also, should we be able to stop all pollling for a given network? (perhaps by omitting the polling token and only giving the networkCLientId?)

Also, should we be able to stop polling for a token, and not require passing in the networkCLientId?

File: packages/gas-fee-controller/src/GasFeeControllerPolling.ts
Comment (BelfordZ) [inline packages/gas-fee-controller/src/GasFeeControllerPolling.ts:13]: Should this be added directly to the base controller instead?

File: packages/gas-fee-controller/src/GasFeeControllerPolling.ts
Comment (BelfordZ) [inline packages/gas-fee-controller/src/GasFeeControllerPolling.ts:91]: Should each networkClientId get its own polling / setInterval? Or should we have 1 single interval, each time it occurs we call each of the `executePoll` methods?

Comment (adonesky1) [reply packages/gas-fee-controller/src/GasFeeControllerPolling.ts:91]: While I see that would be a slightly cleaner to not have another collection per networkClientId, we were thinking it might actually be better for performance to have the network requests staggered.

File: packages/gas-fee-controller/src/GasFeeControllerPolling.ts
Comment (adonesky1) [reply packages/gas-fee-controller/src/GasFeeControllerPolling.ts:13]: We wanted to separate this out to separate concerns and perhaps this could be the form the polling abstraction takes for other controllers to inherit this functionality.

File: packages/gas-fee-controller/src/GasFeeControllerPolling.ts
Comment (jiexi) [reply packages/gas-fee-controller/src/GasFeeControllerPolling.ts:91]: I like @BelfordZ 's thinking here of keeping everything on the same "heartbeat". I'm not saying we do this, but we could use the same interval + random wait.

This also reminds me, is setInterval the right approach, or would we be using setTimeout?

**Comment by Gudahtt** [comment]
Generally this looks good! Will give it a more detailed review later

File: packages/gas-fee-controller/src/GasFeeControllerPolling.ts
Comment (adonesky1) [reply packages/gas-fee-controller/src/GasFeeControllerPolling.ts:91]: I suppose it could worth testing the performance tradeoffs. What, if anything, is the basis for this preference other than a cleaner abstraction?

Comment (Gudahtt) [reply packages/gas-fee-controller/src/GasFeeControllerPolling.ts:91]: Might be better to use `setTimeout` over `setInterval to avoid overlapping queries

See https://developer.mozilla.org/en-US/docs/Web/API/setInterval#ensure_that_execution_duration_is_shorter_than_interval_frequency

File: packages/gas-fee-controller/src/GasFeeControllerPolling.ts
Comment (Gudahtt) [reply packages/gas-fee-controller/src/GasFeeControllerPolling.ts:13]: Yeah, this lets us add polling only to controllers with polling.

Using inheritance like this doesn't seem _ideal_. It wouldn't compose well if we have another "controller modification" type interface that we want to add on as well later. Better to use an interface if possible, which we can compose multiple of. But we can cross that bridge when we come to it

With an interface, the one thing I'm not sure of offhand is we get access to `this.messenger`.

File: packages/gas-fee-controller/src/GasFeeControllerPolling.ts
Comment (adonesky1) [reply packages/gas-fee-controller/src/GasFeeControllerPolling.ts]: done here: https://github.com/MetaMask/core/pull/1673/commits/5317aca24f4acee8bffe8f9c1f695592d4a1566b

File: packages/gas-fee-controller/src/GasFeeControllerPolling.ts
Comment (adonesky1) [reply packages/gas-fee-controller/src/GasFeeControllerPolling.ts:13]: Agree, greater composability would be ideal, but if we used an interface we would have to duplicate the start, stop, stopAll and poll implementations in each controller though?

Comment (Gudahtt) [reply packages/gas-fee-controller/src/GasFeeControllerPolling.ts:13]: Yes, but that's where we could use a library. We could move those implementations to `controller-utils` to reuse them, and use an interface for establishing a common interface for dealing with polling controllers

Comment (adonesky1) [reply packages/gas-fee-controller/src/GasFeeControllerPolling.ts:13]: Yep that makes sense

Comment (shanejonas) [reply packages/gas-fee-controller/src/GasFeeControllerPolling.ts:12]: could maybe use mixins pattern that typescript supports for this: https://www.typescriptlang.org/play#example/mixins

File: packages/controller-utils/package.json
Comment (adonesky1) [inline packages/controller-utils/package.json:33]: We can remove these now

File: packages/controller-utils/package.json
Comment (adonesky1) [inline packages/controller-utils/package.json:41]: these too

**Comment by Gudahtt** [comment]
I suspect that CI will not let us use the new package in the gas fee controller package until we merge and release it separately

File: packages/gas-fee-controller/src/GasFeeControllerPolling.ts
Comment (Gudahtt) [reply packages/gas-fee-controller/src/GasFeeControllerPolling.ts]: Yeah, the mixin pattern looks applicable here. It seems that class decorators in TypeScript v5 and above might offer a more concise syntax for using the mixin pattern as well. But in either case, starting with an abstract class seems reasonable to me. It looks like even if we start using a mixin, the mixin could use this abstract class.

**Review by Gudahtt — COMMENTED** [review]
Looks great! Maybe we can move the new package into a separate PR and get that merged and released to unblock the gas fee controller changes.

File: packages/polling-controller/src/PollingController.ts
Comment (adonesky1) [reply packages/polling-controller/src/PollingController.ts:23]: and probably shouldn't a hardcoded value?

File: packages/gas-fee-controller/src/GasFeeController.ts
Comment (shanejonas) [inline packages/gas-fee-controller/src/GasFeeController.ts:442]:
```suggestion
```

File: packages/gas-fee-controller/src/GasFeeController.ts
Comment (shanejonas) [inline packages/gas-fee-controller/src/GasFeeController.ts:416]:
```suggestion
```

File: packages/gas-fee-controller/src/GasFeeController.ts
Comment (jiexi) [inline packages/gas-fee-controller/src/GasFeeController.ts:398]: this call is synchronous i think

File: packages/gas-fee-controller/src/GasFeeController.ts
Comment (jiexi) [inline packages/gas-fee-controller/src/GasFeeController.ts:199]:
```suggestion
export type ChainGasFeeState =
```

?

**Review by jiexi — APPROVED** [review]

File: packages/network-controller/src/NetworkController.ts
Comment (shanejonas) [inline packages/network-controller/src/NetworkController.ts:1011]: 😂

File: packages/gas-fee-controller/src/GasFeeController.ts
Comment (adonesky1) [reply packages/gas-fee-controller/src/GasFeeController.ts:199]: Hmmm I agree `GasFeeStateOld` isn't great but I'm not sure I understand why `ChainGasFeeState` makes sense? Maybe `SingleChainGasFeeState`?

Comment (adonesky1) [reply packages/gas-fee-controller/src/GasFeeController.ts]: Changed to `SingleChainGasFeeState` here: https://github.com/MetaMask/core/pull/1673/commits/9b53d8c39a968c10d089881357c9ce0f3dd63152

Comment (adonesky1) [reply packages/gas-fee-controller/src/GasFeeController.ts]: Open to other suggestions though

**Review by shanejonas — APPROVED** [review]

**Pull request merged**
